training/CHAPEL_TUTORIAL.md

```diff
@@ -993,358 +1072,618 @@ Expected Output:
-|      2 |     14 |     20 |         20 |             0 |             26 |               14 |               20 |
-|      3 |     10 |     20 |         20 |             0 |             30 |               10 |               20 |
-```
-
-<a id="ex25"></a>
-### Example 25: Common Pitfalls
-In this section we will cover some common pitfalls that can hurt performance.
-  - Using ranges as leading iterators of parallel loops with distributed arrays (i.e. `forall (i, a) in zip(0..#A.size, A) {`).
-    - Ranges aren't distributed, so this actually turns into a parallel loop that only executes on locale 0.
-    So a task per core on locale 0 will be executing, but the rest of the machine will be unused.
-  - Not using aggregation (with random indices) or bulk transfer (with contiguous indices) for copying.
-    - We saw in the previous section what a big difference aggregation can make when copying random elements
-    out of or into a distributed array.
-    - When we want to copy a contiguous block of indices (such as a slice), it's better to use bulk transfer. i.e. `A[start..#stop] = B;`.
-  - Not using refs for component accesses inside loops.
-    - If you find yourself accessing a component of a class during every iteration of a loop, it's often more efficient
-    to save a reference to that component to avoid fetching it every time.
+The `with` keyword here declares `agg` as a [Task-Private variable](https://chapel-lang.org/docs/language/spec/data-parallelism.html#task-private-variables),
+meaning each task will get its own private variable with the same name.
 
-Let's look at an example based on an [actual discussion](https://github.com/Bears-R-Us/arkouda/pull/2159#discussion_r1113699483)
-in arkouda that hits on a lot of these points. To motivate the use of `ref`s we need to create a simplified version of
-an arkouda `SymEntry`. All you really need to know is it's a class that contains a `.a` property which is a distributed array.
+Let's look at an example, where we copy values from one distributed array into another but shifted over by 3.
+We'll call the array we are copying from `src` and the array we are copying to `dst`.
+This example is designed so when ran with 2 locales, every value will be copied to or from a remote location.
 
-In this example we will take `hashes`, a distributed array contain a tuple of uints `(uint, uint)`. And copy the values
-into `upper` and `lower`, 2 distributed `uint` arrays.
-```Chapel
-use BlockDist, CommDiagnostics, Time;
+When you do a `forall` loop over a distributed array, each locale will handle the values that are local to it.
+So we can go about this 2 different ways:
+* We could loop over the `src` array, so the source values will always be local. We then copy our local values
+into a remote position of the `dst` array using a `DstAggregator`
+* Or we loop over the `dst` array, so destination values are local. We then want to copy remote values from
```

Comment:
You definitely state them here, but it might be worth regrouping them with where they are defined up above and/or adding the more detailed `x` on `locale_i` and 'y` on `local_j` example you previously used.

Reply:
tbh I could prob just copy this section and say "in the following example ..."

Reply:
yah! we could def move this up

Reply:
we chatted offline and decided to leave this bit where it is. Instead we decided to add an example write below the definitions that's structured similarly to the one used to define `remote` and `local`